pkg/render/csi.go

```diff
@@ -0,0 +1,338 @@
+			corev1.EnvVar{
+				Name:  "ADDRESS",
+				Value: filepath.Clean("/csi/csi.sock"),
+			},
+			corev1.EnvVar{
+				Name: "DRIVER_REG_SOCK_PATH",
+				// This path cannot also reference "/csi" because /csi only exists inside of the pod, but this path
+				// is used by the kubelet on the host node to issue CSI operations
+				Value: filepath.Join(c.cfg.Installation.KubeletVolumePluginPath, "plugins/csi.tigera.io/csi.sock"),
+			},
+			corev1.EnvVar{
+				Name: "KUBE_NODE_NAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "spec.nodeName",
+					},
+				},
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			corev1.VolumeMount{
+				Name:      "socket-dir",
+				MountPath: filepath.Clean("/csi"),
+			},
+			corev1.VolumeMount{
+				Name:      "registration-dir",
+				MountPath: filepath.Clean("/registration"),
+			},
+		},
+	}
+
+	return []corev1.Container{
+		csiContainer,
+		registrarContainer,
+	}
+}
+
+func (c *csiComponent) csiVolumes() []corev1.Volume {
+	hostPathTypeDir := corev1.HostPathDirectory
+	hostPathTypeDirOrCreate := corev1.HostPathDirectoryOrCreate
+	return []corev1.Volume{
+		corev1.Volume{
+			Name: "varrun",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: filepath.Clean("/var/run"),
+				},
+			},
+		},
+		corev1.Volume{
+			Name: "etccalico",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: filepath.Clean("/etc/calico"),
+				},
+			},
+		},
+		corev1.Volume{
+			Name: "kubelet-dir",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: c.cfg.Installation.KubeletVolumePluginPath,
+					Type: &hostPathTypeDir,
+				},
+			},
+		},
+		corev1.Volume{
+			Name: "socket-dir",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: filepath.Join(c.cfg.Installation.KubeletVolumePluginPath, "plugins/csi.tigera.io"),
+					Type: &hostPathTypeDirOrCreate,
+				},
+			},
+		},
+		corev1.Volume{
+			Name: "registration-dir",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: filepath.Join(c.cfg.Installation.KubeletVolumePluginPath, "plugins_registry"),
+					Type: &hostPathTypeDir,
+				},
+			},
+		},
+	}
+}
+
+func (c *csiComponent) csiTemplate() corev1.PodTemplateSpec {
+	templateLabels := map[string]string{
+		"name": CSIDaemonSetName,
+	}
+	templateMeta := metav1.ObjectMeta{
+		Labels: templateLabels,
+	}
+	templateSpec := corev1.PodSpec{
+		Tolerations: c.csiTolerations(),
+		Containers:  c.csiContainers(),
+		Volumes:     c.csiVolumes(),
+	}
+	return corev1.PodTemplateSpec{
+		ObjectMeta: templateMeta,
+		Spec:       templateSpec,
+	}
+}
+
+// csiDaemonset creates the daemonset necessary to enable the CSI driver
+func (c *csiComponent) csiDaemonset() *appsv1.DaemonSet {
+	dsLabels := map[string]string{
+		"app.kubernetes.io/name": CSIDaemonSetName,
```

Comment:
We don't need to specify this because the CreateOrUpdateOrDelete function will take care adding this, see [this for where it happens](https://github.com/tigera/operator/blob/93ea11c32e182aea91a49aeae56529895af8968e/pkg/controller/utils/component.go#L487)

Reply:
I wasn't aware of this piece of infra, thanks for the pointer.